cpp/src/join/sort_join.cuh

```diff
@@ -1,247 +1,325 @@
-*/
-
-/* Sort-based join using moderngpu */
+ * Copyright (c) 2017, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+/* Sort-based join using thrust */
 
 #include "cudf/functions.h"
 #include "cudf/types.h"
+#include "join_types.h"
+#include "full_join.cuh"
 
 #include <utility>
-#include <moderngpu/kernel_sortedsearch.hxx>
-#include <moderngpu/kernel_scan.hxx>
-#include <moderngpu/kernel_load_balance.hxx>
-
-using namespace mgpu;
+#include <thrust/transform_scan.h>
 
-struct _join_bounds {
-    mem_t<int> lower, upper;
+template <typename index_type>
+struct JoinBounds {
+    thrust::device_vector<index_type> lower;
+    thrust::device_vector<index_type> upper;
 };
 
-template<typename launch_arg_t = empty_t,
-  typename a_it, typename b_it, typename comp_t>
-_join_bounds compute_join_bounds(a_it a, int a_count, b_it b, int b_count,
-    comp_t comp, context_t& context) {
-
-    mem_t<int> lower(a_count, context);
-    mem_t<int> upper(a_count, context);
-    sorted_search<bounds_lower, launch_arg_t>(a, a_count, b, b_count,
-    lower.data(), comp, context);
-    sorted_search<bounds_upper, launch_arg_t>(a, a_count, b, b_count,
-    upper.data(), comp, context);
-
-    // Prepare output
-    _join_bounds bounds;
-    lower.swap(bounds.lower);
-    upper.swap(bounds.upper);
+/* --------------------------------------------------------------------------*/
+/**
+ * @Synopsis  Computes the lower and upper bound searches of left column values
+ * in the right column
+ *
+ * @Param l The left column to be joined
+ * @Param l_count The size of the left column
+ * @Param r The right column to be joined
+ * @Param r_count The size of the right column
+ * @tparam T Type of column data to be joined
+ * @tparam index_type Output type for the index calculation
+ * @tparam size_type Type for size specification of the columns
+ *
+ * @Returns JoinBounds struct containing the lower and upper bound search results
+ */
+/* ----------------------------------------------------------------------------*/
+template<typename T,
+    typename index_type,
+    typename size_type>
+JoinBounds<index_type>
+compute_join_bounds(
+        T const * const l, size_type l_count,
+        T const * const r, size_type r_count) {
+    JoinBounds<index_type> bounds;
+    bounds.lower.resize(l_count);
+    bounds.upper.resize(l_count);
+    thrust::lower_bound(
+            thrust::device,
```

Comment:
Please use an RMM allocator for these Thrust algorithms to optimize intermediate allocations.